### PR TITLE
373 generate AI resume with DB interaction

### PR DESCRIPTION
**Pull request by ThunderIW**
## 📝 Description

Adds a new AI-powered resume generator (Version 2) that uses LangChain and Google Gemini to generate polished, professional resume entries from analyzed project data stored in the database.

Key components:
- `ResumeProjectInfo` dataclass: Extracts and structures project data from the database
- `AIResumeEntry` dataclass: Represents the AI-generated resume output
- `GenerateResumeAI_Ver2` class: Orchestrates fetching project data and invoking the Gemini LLM to generate action-oriented, impactful resume entries

**Closes:** #373,

---

## 📦 Why This PR Is Large                                                                                                                                                                                                        
                                                                                                                                                                                                                                    
  This PR appears large due to several factors beyond the core feature implementation:                                                                                                                                              
                                                                                                                                                                                                                                    
  1. **Docstrings Added:** 8 comprehensive docstrings were added throughout the codebase (2 dataclass docstrings, 1 class method docstring for `from_project_data`, and 5 method docstrings for the main generator class), including
   detailed Args/Returns documentation.                                                                                                                                                                                             
  2. **New Feature File**: The main implementation (`Generate_Resume_AI_Ver2.py`) includes 2 dataclasses, 1 generator class with LangChain integration, and a detailed LLM prompt template.
  3. **Test File Added**: `test_Generate_Resume_AI_Ver2.py` contains 21 tests across 3 classes covering dataclass parsing (including edge cases for string/missing `project_type`), database interaction, context building, and LLM
  chain behavior.
  4. **Merge Commits:** Multiple merges from development and Timstamp-Primary-Key branches were required to keep this feature branch up to date, adding diff noise from upstream changes and ensure that the database system is up
  to date


  Actual code for this feature: Generate_Resume_AI_Ver2.py contains only ~145 lines of functional code (excluding docstrings, comments, and blank lines). The test file adds ~190 functional lines across 21 test cases.

  | File | Change | Total Lines | Functional Lines |
  |------|--------|-------------|------------------|
  | `src/reporting/Generate_Resume_AI_Ver2.py` | New — AI resume generator with LangChain/Gemini integration | 301 | ~145 |
  | `test/test_Generate_Resume_AI_Ver2.py` | New — 21 tests across 3 classes with mocked DB/LLM | 234 | ~190 |


---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

Comprehensive unit tests added with mocked database and LLM calls.


### How to Run Tests

**Local (pytest):**
```bash
pytest test/test_Generate_Resume_AI_Ver2.py -v
```
**Docker:**
```bash
docker-compose run --rm app pytest test/test_Generate_Resume_AI_Ver2.py -v
```



| Test Class | Test Name | Description | Status |
|------------|-----------|-------------|:------:|
| `TestResumeProjectInfo` | `test_populates_all_fields` | Verifies all fields are correctly extracted from project data | ✅ |
| `TestResumeProjectInfo` | `test_empty_dict_returns_defaults` | Ensures default values when given empty input | ✅ |
| `TestResumeProjectInfo` | `test_missing_oop_analysis_defaults` | Handles missing OOP analysis gracefully | ✅ |
| `TestAIResumeEntry` | `test_all_fields_stored` | Verifies dataclass stores all provided fields | ✅ |
| `TestAIResumeEntry` | `test_optional_defaults` | Confirms optional fields default correctly | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_init_stores_name_and_checks_db` | Constructor stores project name and checks database | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_init_project_not_found` | Handles non-existent projects | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_init_no_api_key_required` | API key not required until generation | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_get_info_returns_dataclass` | Returns properly populated ResumeProjectInfo | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_get_info_returns_none_when_missing` | Returns None when project data missing | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_build_context_contains_fields` | Context string includes all relevant project fields | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_build_context_empty_when_no_data` | Returns empty string when no data available | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_build_context_not_detected_for_empty_lists` | Shows "Not detected" for empty language/framework lists | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_generate_returns_none_project_missing` | Returns None for missing project | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_generate_returns_none_no_data` | Returns None when no analysis data | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_generate_returns_entry_on_success` | Returns AIResumeEntry with correct fields on success | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_generate_passes_context_to_chain` | Verifies context is passed to LLM chain | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_chain_cached_after_first_call` | LLM chain is lazily initialized and cached | ✅ |
| `TestGenerateResumeAI_Ver2` | `test_chain_raises_without_api_key` | Raises RuntimeError when GOOGLE_API_KEY missing | ✅ |




### Local pytest result
<img width="1821" height="559" alt="image" src="https://github.com/user-attachments/assets/5f73199d-a444-4d63-9238-5b092a397537" />

### Docker pytest result
<img width="1819" height="250" alt="image" src="https://github.com/user-attachments/assets/dcf87485-ab5d-4c1b-a9f8-34aa8a922c9a" />




---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [ ] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> N/A - Backend feature with no UI changes

<details>
<summary>Click to expand screenshots</summary>

No screenshots - this is a backend AI service module.

</details>

### CODE REVIEW

File: src/reporting/Generate_Resume_AI_Ver2.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+                to contain top-level keys 'resume_item', 'oop_analysis',
+                'project_type', and 'duration_estimate'.
+
+        Returns:
+            ResumeProjectInfo: A populated dataclass with project name, type,
+                skills, languages, frameworks, summary, highlights, OOP score
+                and rating, duration estimate, and framework sources.
+        """
+        resume_item = data.get("resume_item", {})
+        oop_analysis = data.get("oop_analysis", {})
```
Comment (Puneet-Maan) [inline src/reporting/Generate_Resume_AI_Ver2.py]: OOP score handling assumes score is a dict, if it’s just a number, calling .get() crashes.

**Comment by mahigangal** [comment]
Crash fix is in place, safe handling for dict/string/missing project_type. A new test added, which is passing. approved

File: src/reporting/Generate_Resume_AI_Ver2.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        if self._chain is None:
+            google_api_key = os.getenv("GOOGLE_API_KEY")
+            if not google_api_key:
+                raise RuntimeError("Missing GOOGLE_API_KEY in .env file")
+
+            llm = ChatGoogleGenerativeAI(
+                model="gemini-2.0-flash",
+                google_api_key=google_api_key
+            )
+            parser = JsonOutputParser()
```
Comment (Puneet-Maan) [inline src/reporting/Generate_Resume_AI_Ver2.py]: add “Return ONLY valid JSON, no code fences or extra text” to reduce parser failures.

Comment (ThunderIW) [reply src/reporting/Generate_Resume_AI_Ver2.py]: > add “Return ONLY valid JSON, no code fences or extra text” to reduce parser failures.

Thank you for the suggestion good move for this

**Review by Puneet-Maan — APPROVED** [review]
requested changes made, ty @ThunderIW . approving
